File: LeetCode June Challenge/CourseSchedule3.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int scheduleCourse(vector<vector<int>>& courses) {
        sort(courses.begin(), courses.end(), [](auto &a, auto &b){return a[1] < b[1];});
        priority_queue<int> maxheap;
        int time = 0;
        for(auto course : courses){
            if(time + course[0] <= course[1]){
                time += course[0];
                maxheap.push(course[0]);
            }
            else if(maxheap.size() && maxheap.top() > course[0]){
                time += course[0] - maxheap.top();
                maxheap.pop();
                maxheap.push(course[0]);
            }
        }
        return maxheap.size();        
    }
};
int main(){
    vector<vector<int>> courses = {{100,200},{200,1300},{1000,1250},{2000,3200}};
    Solution s;
    cout<<s.scheduleCourse(courses)<<endl;
    return 0;
}

File: Year-2022/LeetCode June Challenge/CourseSchedule3.java
import java.util.*;
class Solution {
    public int scheduleCourse(int[][] courses) {
        Arrays.sort(courses, (a, b) -> a[1] - b[1]);
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        int time = 0;
        for(int[] course : courses){
            if(course[0] + time <= course[1]){
                time += course[0];
                maxHeap.add(course[0]);
            }
            else if(maxHeap.size() != 0 && maxHeap.peek() > course[0]){
                time += course[0] - maxHeap.poll();
                maxHeap.add(course[0]);
            }
        }
        return maxHeap.size();
    }
}
public class CourseSchedule3 {
    public static void main(String[] args) {
        int[][] courses = {{100,200},{200,1300},{1000,1250},{2000,3200}};
        Solution s = new Solution();
        System.err.println(s.scheduleCourse(courses));
    }
}
